### ISSUE TITLE
How to do something like beforeEach/afterEach for each test case (cucumber js integration)?

### ISSUE THREAD

**Issue by rgeraldporter** [mucsi96/nightwatch-api#546]
Hi, I'm trying to figure out how to "reset" the browser (clear local storage, etc) upon either the completion of each test case, or at the start of each one. However, `Before` & `After` run per-scenario, and `BeforeAll` & `AfterAll` run before and after the entire test run. There seems to be no API that I can find that covers this situation. There was once a proposed `BeforeFeature` hook to cucumber-js but it was never added.

Any ideas are welcomed.

**Comment by DmytroShtanko1** [comment]
Hey! You can write universe beforeEach and other in hooks file or global use scenario as input data. You can find needed annotations (@) there and write your conditions. Hope it help!

**Comment by mucsi96** [comment]
Before or After should be fine as they run before/after every scenario. Isn't that what you need?

**Comment by rgeraldporter** [comment]
@mucsi96 when I tried that it did something between each 

```
Scenario: Do some thing
```

rather than per feature file.

**Comment by mucsi96** [comment]
Sorry I don't get what you are saying. Feature file is not a test case. Feature file is a test suite. Why you want to run the session restoration per test suite rather the test case (scenario). Not sure why Before or After not fitting your use case.

**Comment by rgeraldporter** [comment]
@mucsi96 Sorry about the language confusion. In reading some forum answers elsewhere it seemed the term "test case" was equivalent to a feature file.

Each feature file (test suite) in my case logs in to an application, does its thing, then logs out.

If a failure occurs in a test suite that causes the UI to get stuck somewhere that cannot access the "log out" button, all future test suites fail because they can't log in (because the browser never logged out, there is no "Log in" button).

What would solve this if in a, lets call it, `BeforeFeature` a new browser session started, then in an `AfterFeature` that browser session ended, this would solve the problem (clearing local storage, etc).

I hope this clears it up.

**Comment by mucsi96** [comment]
No problem :) The use case is clear. I think there is a confusing how cucumber works maybe. So let's imagine a feature file which contains some scenarios which require the user to be logged in. You can use some background steps in the feature file to log in the user. Even nicer would be to create a tag e.g. requiresLogin and create a before hook matching this tag and doing the login. Whichever way you go the Before hook will be executed before login steps. So you can do a cleanup. Like reload the page clear session and localstorage or delete cookies. It might be confusing but background steps of feature file are not running before the feature file. Feature file itself is not running at all. It's just a holder for scenarios and scenarios are running. In case a feature background the background steps are running as part of every scenario in the beginning. 

To sum up. Try to create a Before hook with session cleanup and page reload. It should work fine

**Comment by rgeraldporter** [comment]
@mucsi96 Thanks, I'll give that a try. This is further complicated though, by each feature file specifying _which_ user logs in, and may log in or out multiple times to test administrator changes and non-administrator access levels. There's really no way for me to have one user who logs in/out for every feature file as it stands due to the diverse cases. 

I was pretty sure `Before` was running on each `Scenario:` yesterday because when I put a create/end session in there the second `scenario` would fail but I'll give it a go again.

**Comment by mucsi96** [comment]
Would be create if you create a very simple example repo showing the issue.

Not forget that you need to do the login in every scenario. Scenarios should be isolated and not depend on each other. Also there order in feature file should not matter.

**Comment by rgeraldporter** [comment]
Our test cases often involve complex processes / tasks, such as building a form (with all kinds of variants of what elements could be in said form), making sure it renders properly from a non-admin perspective, and ensuring data that is submitted is available in a forms report section. Our scenarios would be incredibly long if they had to be completely isolated.

In the example here: https://cucumber.io/docs/gherkin/reference/

I interpreted that the "Given" in the second scenario requires the first scenario to run in order to be acted upon. Is that not the intended case in the JS version of Cucumber?

**Comment by mucsi96** [comment]
I think you interpreted the example not correctly. Cucumber scenarios should not belong on each other or act upon. Even if you have complex things to test you need to have steps that have more complex step definitions and keep the scenarios not very long and isolated. In your can can have many different steps which fills the entire forms.

**Comment by mucsi96** [comment]
But the idea of Cucumber scenarios is that they are isolated

**Comment by mucsi96** [comment]
Have you seen this one https://cucumber.io/docs/cucumber/state/ ?

**Comment by rgeraldporter** [comment]
I see that some documentation mentions it should be isolated. It's hard to hear that now after writing dozens of feature files. If that's the case, why do the feature files always run in a linear fashion? I feel we're not going to be the only ones who interpret things this way, especially those who have very complex software processes that need to happen to test certain components.

I looked at the state link but in an E2E we do not have access to the database to clear it, and some of our database records are in a CQRS situation so data can't be "cleared" as it would break the audit trail.

It may be we'll just have to live with not being able to reset the browser, and just fix the bugs that cause the login session to persist, which we need to fix anyways. We may just have to do a clean-room approach where new tests are less state-dependent.

Thanks for the assistance in understand this!
